Kotlin/lib/src/main/java/com/wolt/blurhashkt/BlurHashDecoder.kt

```diff
@@ -77,27 +88,85 @@ object BlurHashDecoder {
+            useCache: Boolean
     ): Bitmap {
-        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
+        // use an array for better performance when writing pixel colors
+        val imageArray = IntArray(width * height)
+        val calculateCosX = !useCache || !cacheCosinesX.containsKey(width * numCompX)
+        val cosinesX = getCosinesX(calculateCosX, width, numCompX)
+        val calculateCosY = !useCache || !cacheCosinesY.containsKey(height * numCompY)
+        val cosinesY = getCosinesY(calculateCosY, height, numCompY)
         for (y in 0 until height) {
             for (x in 0 until width) {
                 var r = 0f
                 var g = 0f
                 var b = 0f
                 for (j in 0 until numCompY) {
                     for (i in 0 until numCompX) {
-                        val basis = (cos(PI * x * i / width) * cos(PI * y * j / height)).toFloat()
+                        val cosX = getCosX(calculateCosX, cosinesX, i, numCompX, x, width)
+                        val cosY = getCosY(calculateCosY, cosinesY, j, numCompY, y, height)
+                        val basis = (cosX * cosY).toFloat()
                         val color = colors[j * numCompX + i]
                         r += color[0] * basis
                         g += color[1] * basis
                         b += color[2] * basis
                     }
                 }
-                bitmap.setPixel(x, y, Color.rgb(linearToSrgb(r), linearToSrgb(g), linearToSrgb(b)))
+                imageArray[x + width * y] = Color.rgb(linearToSrgb(r), linearToSrgb(g), linearToSrgb(b))
+            }
+        }
+        return Bitmap.createBitmap(imageArray, width, height, Bitmap.Config.ARGB_8888)
+    }
+
+    private fun getCosinesY(calculateCosY: Boolean, height: Int, numCompY: Int): DoubleArray {
+        val cosinesY: DoubleArray
+        if (calculateCosY) {
+            cosinesY = DoubleArray(height * numCompY)
+            cacheCosinesY[height * numCompY] = cosinesY
+        } else {
+            cosinesY = cacheCosinesY[height * numCompY]!!
+        }
+        return cosinesY
+    }
+
+    private fun getCosY(
+            calculateCosY: Boolean,
+            cosinesY: DoubleArray,
+            j: Int,
+            numCompY: Int,
+            y: Int,
+            height: Int
+    ): Double {
+        if (calculateCosY) {
+            cosinesY[j + numCompY * y] = cos(Math.PI * y * j / height)
+        }
+        return cosinesY[j + numCompY * y]
+    }
+
+    private fun getCosX(
+            calculateCosX: Boolean,
+            cosinesX: DoubleArray,
+            i: Int,
+            numCompX: Int,
+            x: Int,
+            width: Int
+    ): Double {
+        if (calculateCosX) {
+            cosinesX[i + numCompX * x] = cos(Math.PI * x * i / width)
+        }
+        return cosinesX[i + numCompX * x]
+    }
+
+    private fun getCosinesX(calculateCosX: Boolean, width: Int, numCompX: Int): DoubleArray {
```

Comment:
Understanding that this is a get function for val `cosinesX`, but `getCosX` and `getCosinesX` is quite confusing first time I try digging through the code. Same with `getCosinesY` and `getCosY`.

Reply:
you are right, it is confusing, will change the names 👍